File: dotnet-docs-viewer/Services/PreviewService.cs
using System.Transactions;
using Common.Utils;
using DB.Model;
using Microsoft.AspNetCore.StaticFiles;
using Microsoft.Extensions.Options;
using WebApp.Dto;
using WebApp.Repository;
using WebApp.Types;

namespace WebApp.Service
{
    public interface IPreviewService
    {
        Task<byte[]> FindFileAsync(FileResponse fileResponse);
        Task SaveFileAsync(int attachFile, byte[] image, string targetFolder, int index);
        Task SaveFileAsync(int attachFile, FileResponse fileResponse);
    }
    
    public class PreviewService
    {
        private readonly IRepository<FileInfoBaseModel> _fileRepository;
        private readonly IRepository<PreviewInfoBaseModel> _previewRepository;
        private readonly IOptions<FileSetting> _fileSetting;

        private const string FileDirectory = "cmmn";

        public PreviewService(IRepository<FileInfoBaseModel> fileRepository, IRepository<PreviewInfoBaseModel> previewRepository, IOptions<FileSetting> fileSetting)
        {
            _fileRepository = fileRepository;
            _previewRepository = previewRepository;
            _fileSetting = fileSetting;
        }

        // ───────────────────────────────────────────────────────────
        // 파일 변환 후 바로 반환하는 기능
        // ───────────────────────────────────────────────────────────
        public async Task<byte[]> FindFileAsync(FileResponse fileResponse)
        {
            byte[] resource = fileResponse.Resource;
            string mimeType = fileResponse.MimeType;

            var previewType = GetPreviewType(mimeType);

            return previewType switch
            {
                PreviewType.PPTX => await ConvertPptxToImage(resource),
                PreviewType.XLSX => await ConvertXlsxToImage(resource),
                PreviewType.PDF => await ConvertPdfToImage(resource),
                _ => resource
            };
        }

        // ---------------------------------------------------------------------
        // 파일 변환 후 이미지 여러 개 생성하여 DB에 저장하는 로직
        // ---------------------------------------------------------------------
        public async Task SaveFileAsync(int attachFile, byte[] image, string targetFolder, int index)
        {
            using var scope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled);

            var fileInfo = await FindFileByIdAsync(attachFile);

            string rootPath = _fileSetting.Value.ImagePath;
            string uploadDirectory = FileUtil.GetDirectory(rootPath, targetFolder);
            string uploadPath = FileUtil.GetFilePath(uploadDirectory);

            string mimeType = FindMimeType(image);
            int fileSize = image.Length;

            var previewInfo = PreviewInfoModel.Of(fileInfo, uploadPath, mimeType, fileSize);

            await _previewRepository.SaveAsync(previewInfo);

            // 실제 파일 저장 (암호화)
            CryptoUtil.EncryptFile(image, FileUtil.GetTargetFile(uploadPath));

            scope.Complete();
        }

        // Overload (Docx/Pdf/Xlsx/Pptx → 여러 개 pages 저장)
        public async Task SaveFileAsync(int attachFile, FileResponse fileResponse)
        {
            byte[] resource = fileResponse.Resource;
            string mimeType = fileResponse.MimeType;

            var previewType = GetPreviewType(mimeType);

            switch (previewType)
            {
                case PreviewType.PPTX:
                    await CreatePptxToImage(attachFile, resource);
                    break;
                case PreviewType.XLSX:
                    await CreateXlsxToImage(attachFile, resource);
                    break;
                case PreviewType.PDF:
                    await CreatePdfToImage(attachFile, resource);
                    break;
            }
        }

        // ───────────────────────────────────────────────────────────
        // 유틸 메서드
        // ───────────────────────────────────────────────────────────
        private async Task<FileInfoBaseModel> FindFileByIdAsync(int id)
        {
            var file = await _fileRepository.GetByIdAsync(id);

            return file ?? throw new FileNotFoundException("존재하지 않는 파일입니다.");
        }

        private string FindMimeType(byte[] bytes)
        {
            new FileExtensionContentTypeProvider().TryGetContentType("file", out var mime);
            return mime ?? "application/octet-stream";
        }

        private PreviewType GetPreviewType(string mimeType)
        {
            if (FileTypeUtil.IsPptx(mimeType)) return PreviewType.PPTX;
            if (FileTypeUtil.IsXlsx(mimeType)) return PreviewType.XLSX;
            if (FileTypeUtil.IsDocx(mimeType)) return PreviewType.DOCX;
            if (FileTypeUtil.IsPdf(mimeType)) return PreviewType.PDF;

            return PreviewType.NONE;
        }

        // ───────────────────────────────────────────────────────────
        // 변환 기능 (DocumentUtil 사용)
        // ───────────────────────────────────────────────────────────
        private async Task<byte[]> ConvertPdfToImage(byte[] pdf)
        {
            var images = await DocumentUtil.ConvertPdfToPng(pdf);
            return DocumentUtil.MergeImagesVertically(images);
        }

        private async Task CreatePdfToImage(int attachFile, byte[] pdf)
        {
            var images = await DocumentUtil.ConvertPdfToPng(pdf);
            int index = 1;

            foreach (var img in images)
            {
                await SaveFileAsync(attachFile, img, FileDirectory, index++);
            }
        }

        private async Task<byte[]> ConvertPptxToImage(byte[] pptx)
        {
            var images = await DocumentUtil.ConvertPptxToPng(pptx);
            return DocumentUtil.MergeImagesVertically(images);
        }

        private async Task CreatePptxToImage(int attachFile, byte[] pptx)
        {
            var images = await DocumentUtil.ConvertPptxToPng(pptx);
            int index = 1;

            foreach (var img in images)
            {
                await SaveFileAsync(attachFile, img, FileDirectory, index++);
            }
        }

        private async Task<byte[]> ConvertXlsxToImage(byte[] xlsx)
        {
            var images = await DocumentUtil.ConvertXlsxToPng(xlsx);
            return DocumentUtil.MergeImagesVertically(images);
        }

        private async Task CreateXlsxToImage(int attachFile, byte[] xlsx)
        {
            var images = await DocumentUtil.ConvertXlsxToPng(xlsx);
            int index = 1;

            foreach (var img in images)
            {
                await SaveFileAsync(attachFile, img, FileDirectory, index++);
            }
        }
    }
}

File: spring-docs-viewer-libre-office/jpa/src/main/java/com/docs/viewer/common/preview/service/PreviewService.java
package com.docs.viewer.common.preview.service;

import com.docs.viewer.common.file.dto.response.FileResponse;
import com.docs.viewer.common.file.entity.File;
import com.docs.viewer.common.file.repository.FileRepository;
import com.docs.viewer.common.preview.entity.Preview;
import com.docs.viewer.common.preview.repository.PreviewRepository;
import com.docs.viewer.common.preview.type.PreviewType;
import com.docs.viewer.global.common.setting.FileSetting;
import com.docs.viewer.global.common.utils.CryptoUtil;
import com.docs.viewer.global.common.utils.FileTypeUtil;
import com.docs.viewer.global.common.utils.FileUtil;
import com.docs.viewer.global.error.exception.NotFoundException;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.tika.Tika;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.stereotype.Service;

import java.nio.file.Path;

@Slf4j
@Service
@RequiredArgsConstructor
public class PreviewService {

    private static final String FILE_DIRECTORY = "cmmn";

    private final FileRepository fileRepository;
    private final PreviewRepository previewRepository;
    private final FileSetting fileSetting;

    public ByteArrayResource findFile(FileResponse fileResponse) throws Exception {
        ByteArrayResource resource = fileResponse.getResource();
        String mimeType = fileResponse.getMimeType();
        PreviewType previewType = this.getPreviewType(mimeType);
        return switch (previewType) {
            case PPTX -> convertPptxToImage(resource);
            case XLSX -> convertXlsxToImage(resource);
            case DOCX -> convertDocxToImage(resource);
            case PDF -> convertPdfToImage(resource);
            default -> resource;
        };
    }

    @Transactional
    public void saveFile(Integer attachFile, ByteArrayResource image, String targetFolder, int index) throws Exception {
        File fileInfo = this.findFileById(attachFile);
        String rootPath = fileSetting.getImagePath();
        Path uploadDirectory = FileUtil.getUploadDirectory(FileUtil.getDirectory(rootPath, targetFolder));
        String uploadPath = FileUtil.getFilePath(uploadDirectory);
        String mimeType = this.findMimeType(image.getByteArray());
        long fileSize = image.contentLength();
        Preview previewInfo = this.previewRepository.save(Preview.of(fileInfo, uploadPath, mimeType, fileSize));
        CryptoUtil.encryptFile(image.getByteArray(), FileUtil.getTargetFile(previewInfo.getFilePath()));  // 파일 저장
    }

    @Transactional
    public void saveFile(Integer attachFile, FileResponse fileResponse) throws Exception {
        ByteArrayResource resource = fileResponse.getResource();
        String mimeType = fileResponse.getMimeType();
        PreviewType previewType = this.getPreviewType(mimeType);
        switch (previewType) {
            case PPTX -> createPptxToImage(attachFile, resource);
            case XLSX -> createXlsxToImage(attachFile, resource);
            case DOCX -> createDocxToImage(attachFile, resource);
            case PDF -> createPdfToImage(attachFile, resource);
            default -> {
            }
        }
    }

    private File findFileById(Integer attachFile) {
        return this.fileRepository.findById(attachFile.longValue())
                .orElseThrow(() -> new NotFoundException("존재하지 않는 상품코드 입니다."));
    }

    private String findMimeType(byte[] fileBytes) {
        Tika tika = new Tika();
        return tika.detect(fileBytes);
    }

    private PreviewType getPreviewType(String mimeType) {
        if(FileTypeUtil.isPptx(mimeType)) {
            return PreviewType.PPTX;
        } else if (FileTypeUtil.isXlsx(mimeType)) {
            return PreviewType.XLSX;
        } else if (FileTypeUtil.isDocx(mimeType)) {
            return PreviewType.DOCX;
        } else if (FileTypeUtil.isPdf(mimeType)) {
            return PreviewType.PDF;
        } else {
            return PreviewType.NONE;
        }
    }

    private ByteArrayResource convertPdfToImage(ByteArrayResource pdfResource) throws Exception {
        return null;
    }

    private void createPdfToImage(Integer attachFile, ByteArrayResource pdfResource) throws Exception {

    }

    private ByteArrayResource convertDocxToImage(ByteArrayResource docxResource) throws Exception {
        return null;
    }

    private void createDocxToImage(Integer attachFile, ByteArrayResource docxResource) throws Exception {

    }

    private ByteArrayResource convertPptxToImage(ByteArrayResource pptxResource) throws Exception {
        return null;
    }

    private void createPptxToImage(Integer attachFile, ByteArrayResource pptxResource) throws Exception {

    }

    private ByteArrayResource convertXlsxToImage(ByteArrayResource xlsxResource) throws Exception {
        return null;
    }

    private void createXlsxToImage(Integer attachFile, ByteArrayResource xlsxResource) throws Exception {

    }
}
